### ISSUE TITLE
Unable to import events/profiles

### ISSUE THREAD

**Issue by craftyguy** [renyuneyun/Easer#56]
I exported on 0.47 (fdroid version) and tried to import on the latest (0.52 from github), however the file picker won't let me select the exported zip file. This is on Android 8.1.

This screenshot shows that the exports from Easer are not selectable:

![screenshot_20180113-192100](https://user-images.githubusercontent.com/1474209/34912494-f31b0112-f897-11e7-94db-bfa98c62cb3f.png)

**Comment by renyuneyun** [comment]
Hmm... Do you have any other file pickers? Do they behave the same way (i.e. also have this problem)?

Maybe Android 8.0 changed some behaviors in the file manager/selector, but I doubt other apps have followed the change (yet).

**Comment by craftyguy** [comment]
I tried with Amaze, it lets me select the file but Easer doesn't actually import it. There is nothing generated in the easer log either :(

**Comment by renyuneyun** [comment]
Hmm... I figured out where possibly went wrong.
Can you try the import function with a clean Easer (e.g. after "clean data")? I think this is because the data directories already exist so Easer refused to import.

**Comment by craftyguy** [comment]
Hmm, no that did not work. I tried:

1) clear data/cache

2) uninstall/install Easer

3) settings->import

4) used two different pickers (Amaze & Dirs, stock Android 8 picker still doesn't allow selecting .zip export..)

5) after selecting, I am taken to the Easer settings menu again but when I go to Events or Profiles, they are empty.


Is there a directory on /sdcard or something that I need to clear out?

**Comment by renyuneyun** [comment]
No, that's not in `/sdcard`. Just to clarify: the only two scenarios where Easer will write to external storage (`/sdcard`) are: 1. Exporting; 2. Logging.

If you wish, I can provide a patch to (hopefully) solve this problem. Though I'd rather test things more and include this in v0.5.3.

**Comment by craftyguy** [comment]
Hmm, so I cleared out everything that the app creates and it doesn't import. I'd definitely test a patch or the next release, whatever is easiest for you!

**Comment by craftyguy** [comment]
So when you say 'the data directories already exist' is causing this, if I go to (in Android not easer app) 'settings'-->apps-->easer and 'clear data & cache', this should delete all data the app has created under /data... right? Doing that doesn't seem to help the problem, I still cannot import the file.

**Comment by renyuneyun** [comment]
@craftyguy I'm also quite curious why. The only affected lines are:
```
File newFile; // initialized elsewhere; represents the corresponding file in the app storage for one file in the zip
if (!new File(newFile.getParent()).mkdirs()) {
    throw new IOException(String.format("failed to create parent dird for <%s>", newFile.toString()));
}
```
So, I guess directories are also treated as a `ZipEntry`, which makes this problem happen (since creating "the parent directory of a subdirectory in the app storage" is equivalent to creating "the app storage directory", which will always fail [because this directory will always exist when the app runs]).

**Comment by renyuneyun** [comment]
(As a progress indicator:)

Splitting Events to Event Logic + Scenarios has been done and tested.

I'm trying to find a best way to add the control of continuation / persistance / repeatable (as in #55) to Events. After that, the most important change of v0.5.3 will be done and it will be pushed (unless there is anything else important).

**Comment by craftyguy** [comment]
@renyuneyun Ah, your guess makes sense, though I should see that exception printed somewhere if it were true.. right? I don't see it in the Easer log though (under /sdcard), would it be printed to the system log (logcat)?

**Comment by craftyguy** [comment]
@renyuneyun 

Ok, I think this confirms your theory. I re-ran Easer and tried to import the zip with logcat running, here's the exception that was printed to logcat:

```
01-18 16:02:03.261 25601 25601 D pickup  : file
01-18 16:02:03.376 25507 25507 W System.err: java.io.IOException: failed to create parent dird for </data/user/0/ryey.easer/files/event/Charging.xml>
01-18 16:02:03.376 25507 25507 W System.err:    at ryey.easer.core.a.c.a(Unknown Source:69)
01-18 16:02:03.376 25507 25507 W System.err:    at ryey.easer.core.ui.setting.SettingsActivity$a.onActivityResult(Unknown Source:15)
01-18 16:02:03.376 25507 25507 W System.err:    at android.app.Activity.dispatchActivityResult(Activity.java:7293)
01-18 16:02:03.376 25507 25507 W System.err:    at android.app.ActivityThread.deliverResults(ActivityThread.java:4265)
01-18 16:02:03.376 25507 25507 W System.err:    at android.app.ActivityThread.handleSendResult(ActivityThread.java:4313)
01-18 16:02:03.376 25507 25507 W System.err:    at android.app.ActivityThread.-wrap19(Unknown Source:0)
01-18 16:02:03.376 25507 25507 W System.err:    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1645)
01-18 16:02:03.376 25507 25507 W System.err:    at android.os.Handler.dispatchMessage(Handler.java:106)
01-18 16:02:03.376 25507 25507 W System.err:    at android.os.Looper.loop(Looper.java:164)
01-18 16:02:03.376 25507 25507 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:6497)
01-18 16:02:03.376 25507 25507 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
01-18 16:02:03.376 25507 25507 W System.err:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
01-18 16:02:03.376 25507 25507 W System.err:    at com.android.internal.os.ExecInit.main(ExecInit.java:49)
01-18 16:02:03.376 25507 25507 W System.err:    at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
01-18 16:02:03.376 25507 25507 W System.err:    at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:285)

```

**Comment by renyuneyun** [comment]
v0.5.3 is out :)

**Comment by craftyguy** [comment]
Thanks, I will test this right now!

**Comment by craftyguy** [comment]
@renyuneyun 
The apk in this repo for 0.5.3 is broken :(

**Comment by craftyguy** [comment]
@renyuneyun importing my exported profiles worked with v0.5.3, however it locked up when trying to view a profile, and after force-closing it, it now just immediately closes when I try to 'start' the Easer service again.

I will try to provide a log when i get home. Note that I got a warning that my data was 'old' and to covert it, which I did do under settings.

**Comment by renyuneyun** [comment]
Hmm, I don't remember a compability change in these versions (though I was hoping to LOL).

The warning for "old data" is for the old xml format. Have you succeeded in converting the data format (e.g. a toast stating success, or this warning is not shown again on next start up)? Does this problem still exist after converting to the new format?

**Comment by craftyguy** [comment]
Here's the sequence of events:

1) install 0.5.3

2) started easer, then settings->import

3) used Amaze to pick the zip

4) Went to profiles, saw them listed, selected one

5) Easer froze (white screen). Force-closed it

6) Started Easer again, was prompted to convert

7) went to settings->convert data

8) went back to easer main screen and selected 'start' from menu

9) easer 'crashed to desktop' after tapping on 'start'

Subsequent attempt so repeat steps 8-9 have the same effect, Easer closes immediately after tapping on 'start'

**Comment by renyuneyun** [comment]
That seems to be a result of "bad" data. The curious part is that bad data should be ignored when encountered...  (Plus, there shouldn't be any bad data in your situation...)

It would be the best if you can provide a log (either logcat or the logging stored in `/sdcard/logger/` by easer [if you enabled that]).

p.s. If appropriate (e.g. there aren't any sensitive data), you can also provide the exported data (the `.zip` file) so I can debug directly on my side.

**Comment by craftyguy** [comment]
logcat of steps 8 & 9 here: https://craftyguy.net/paste/vLbIYD2q

**Comment by renyuneyun** [comment]
They ought to be fixed in v0.5.3.1 :)

There are three bugs involved in total...

**Comment by craftyguy** [comment]
@renyuneyun looks like it works! I was able to successfully import my exported data and start Easer (after clearing data first, as you suggested). Thank you for fixing this!! :smile: 

The only other issue is with the default picker on Android 8 not recognizing the Easer data export as a valid file to pick. I could make a new issue for this if you want, or keep this one open, or ??
Only reason I suggest keeping that one open is that it seems wrong to require users (presumably with Android 8) to install a file manager app just to be able to use the import feature on Easer. I guess that's ultimately up to you.
